upcoming/composable-pipelines-pattern.md

```diff
@@ -0,0 +1,392 @@
+	# Connect our super-minimal little "workflow" consisting
+	# - a base complementer
+	# - a file reader
+	# - our print statements right here in the loop
+	fa_reader = fasta_reader('chr_y.fa')
+	for line in base_complementer(fa_reader):
+		print line
+````
+
+Have a careful look at this line:
+
+````python
+for line in base_complementer(fa_reader):
+	...
+````
+
+We here see how, rather than sending data into `base_complementer`, we instead send a generator object (`fa_reader`), that `base_complementer` will iterate over, as soon as we start iterating over the the latter.
+
+So when executing the main loop in `main()``, it will step by step drive our little pipeline consisting of three parts; a (FASTA) reader, a base complementer generator function and a print-statement at the end.
+
+Again, the lazy evaluation means that one item at a time will be drawn through the whole pipeline for each iteration in the main loop, without any temporary aggregates (lists or dicts) of lines building up between the components, which means the program needs to use very little memory during the process.
+
+## The Generator pattern in Go
+
+Coming into Go, the author was highly intrigued by all the powerful concurrency patterns made possible in this language, elaborated in blog posts such as [the one on "concurrency patterns"](http://blog.golang.org/pipelines) and ["advanced concurrency patterns"](http://blog.golang.org/advanced-go-concurrency-patterns).
+
+Even more interestingly, the simple and straight-forward generator pattern in python was easy to implement in Go too, as shown in [Rob Pike's Go Concurrency pattern slides](https://talks.golang.org/2012/concurrency.slide#25), and also [listed on this site](http://www.golangpatterns.info/concurrency/generators).
+
+The above python code would read something like this in Go, using the generator pattern (leaving out some imports and const definitions, for brevity):
+
+````go
+
+// ...<snip>...
+
+func fileReaderGen(filename string) chan []byte {
+	fileReadChan := make(chan []byte, BUFSIZE)
+	go func() {
+		file, err := os.Open(filename)
+		if err != nil {
+			log.Fatal(err)
+		} else {
+				scan := bufio.NewScanner(file)
+				for scan.Scan() {
+					// Write to the channel we will return, while copying
+					// the buffer slice (which is re-used) ...
+					fileReadChan <- append([]byte(nil), scan.Bytes()...)
+		}
+		close(fileReadChan)
+		fmt.Println("Closed file reader channel")
+		}
+	}()
+	return fileReadChan
+}
+
+// Translation table, used in the base complementer
+// function below
+var baseConv = [256]byte{
+	'A': 'T',
+	'T': 'A',
+	'C': 'G',
+	'G': 'C',
+	'N': 'N',
+	'\n' : '\n',
+}
+
+func baseComplementGen(inChan chan []byte) chan []byte {
+	returnChan := make(chan []byte, BUFSIZE)
+	go func() {
+		for sequence := range inChan {
+			if sequence[0] != '>' {
+				for pos := range sequence {
+					sequence[pos] = baseConv[sequence[pos]]
+				}
+			}
+			returnChan <- append([]byte(nil), sequence...)
```

Comment:
On 2015-02-10 19:37, Nathan Youngman wrote:
>
> In upcoming/composable-pipelines-pattern.md 
> <https://github.com/gopheracademy/gopheracademy-web/pull/114#discussion_r24437369>:
>
> > +	'C': 'G',
> > +	'G': 'C',
> > +	'N': 'N',
> > +	'\n' : '\n',
> > +}
> > +
> > +func baseComplementGen(inChan chan []byte) chan []byte {
> > +	returnChan := make(chan []byte, BUFSIZE)
> > +	go func() {
> > +		for sequence := range inChan {
> > +			if sequence[0] != '>' {
> > +				for pos := range sequence {
> > +					sequence[pos] = baseConv[sequence[pos]]
> > +				}
> > +			}
> > +			returnChan <- append([]byte(nil), sequence...)
>
> oh. would the copy() built in work?
>

According to this blog [1], copy should work, but will require two lines 
rather than one (one for creating a new slice, and one for the copying), 
while append can be done in one line, and also returns the new array, so 
that we don't need a separate line for the channel send:

Quote:

To fix this problem one can copy the interesting data to a new slice 
before returning it:

func CopyDigits(filename string) []byte {
     b, _ := ioutil.ReadFile(filename)
     b = digitRegexp.Find(b)
     c := make([]byte, len(b))
     copy(c, b)
     return c
}

A more concise version of this function could be constructed by 
using|append|. This is left as an exercise for the reader.

[1] http://blog.golang.org/go-slices-usage-and-internals

I guess one can argue that the copy command is a lot easier to 
understand though, so I could change it if you think it is a better 
recommendation for blog readers?

Best
// Samuel